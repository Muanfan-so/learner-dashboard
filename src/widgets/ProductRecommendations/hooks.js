import { useState, useEffect } from 'react';

import { RequestStates, RequestKeys } from 'data/constants/requests';
import { StrictDict } from 'utils';
import { reduxHooks } from 'hooks';
import { SortKeys } from 'data/constants/app';
import api from './api';
import * as module from './hooks';

export const state = StrictDict({
  requestState: (val) => useState(val), // eslint-disable-line
  data: (val) => useState(val), // eslint-disable-line
});

export const useShowRecommendationsFooter = () => {
  const hasAvailableDashboards = reduxHooks.useHasAvailableDashboards();
  const hasRequestCompleted = reduxHooks.useRequestIsCompleted(RequestKeys.initialize);

  // Hardcoded to not show until experiment related code is implemented
  return {
    shouldShowFooter: false,
    shouldLoadFooter: hasRequestCompleted && !hasAvailableDashboards,
  };
};

export const useMostRecentCourseRunKey = () => {
  const mostRecentCourseRunKey = reduxHooks.useCurrentCourseList({
    sortBy: SortKeys.enrolled,
    filters: [],
    pageSize: 0,
  }).visible[0]?.courseRun?.courseId;

  return mostRecentCourseRunKey;
};

export const useFetchRecommendations = (setRequestState, setData) => {
  const courseRunKey = module.useMostRecentCourseRunKey();

  useEffect(() => {
    let isMounted = true;

    const handleSuccess = (response) => {
      if (isMounted) {
        setData(response.data);
        setRequestState(RequestStates.completed);
      }
    };

    const handleError = () => {
      if (isMounted) {
        setRequestState(RequestStates.failed);
      }
    };

    if (courseRunKey) {
      api
        .fetchCrossProductRecommendations(courseRunKey)
        .then(handleSuccess)
        .catch(handleError);
    } else {
      api
        .fetchAmplitudeRecommendations()
        .then(handleSuccess)
        .catch(handleError);
    }
    return () => {
      isMounted = false;
    };
    /* eslint-disable */
  }, []);
};

export const useProductRecommendationsData = () => {
  const [requestState, setRequestState] = module.state.requestState(RequestStates.pending);
  const [data, setData] = module.state.data({});
  module.useFetchRecommendations(setRequestState, setData);

  return {
    productRecommendations: data,
    isLoading: requestState === RequestStates.pending,
    isLoaded: requestState === RequestStates.completed,
    hasFailed: requestState === RequestStates.failed 
  };
};

export default { useProductRecommendationsData, useShowRecommendationsFooter };
