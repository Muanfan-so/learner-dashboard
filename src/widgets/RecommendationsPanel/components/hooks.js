import { configuration } from 'config';
import { setCookie, getCookie } from 'utils/cookies';

import track from '../track';
import './index.scss';

export const useCourseCardData = (course, isPersonalized) => {
  const handleCourseClick = (e) => {
    e.preventDefault();

    const cookieName = configuration.PERSONALIZED_RECOMMENDATION_COOKIE_NAME;
    let recommendedCourses = getCookie(cookieName);
    if (typeof recommendedCourses === 'undefined') {
      recommendedCourses = { course_keys: [course.courseKey] };
    } else if (!recommendedCourses.course_keys.includes(course.courseKey)) {
      recommendedCourses.course_keys.push(course.courseKey);
    }
    recommendedCourses.is_personalized_recommendation = isPersonalized;
    setCookie(cookieName, JSON.stringify(recommendedCourses), 365);

    track.recommendedCourseClicked(
      course.courseKey,
      isPersonalized,
      course?.marketingUrl,
    )(e);
  };
  return { handleCourseClick };
};

export default useCourseCardData;
