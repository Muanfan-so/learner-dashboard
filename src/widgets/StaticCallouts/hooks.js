import { breakpoints, useWindowSize } from '@edx/paragon';

export const useIsMediumScreen = () => {
  const { width } = useWindowSize();
  return width >= breakpoints.medium.minWidth && width < breakpoints.medium.maxWidth;
};

export default useIsMediumScreen;
