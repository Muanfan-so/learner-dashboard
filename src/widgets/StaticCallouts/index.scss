@import "@edx/paragon/scss/core/core";

.callouts-container {
  display: flex;
  flex-direction: column;
  row-gap: 3rem;
}

.item-badge {
  width: fit-content;
}

.card-description {
  font-size: 0.875rem;
  color: $gray-500;
}

.static-callouts-container {
  display: flex;
  flex-direction: row;
}

.static-callouts-item {
  @include hover() {
    text-decoration: none;
  }
  display: flex;
  flex-direction: column;
  flex-basis: fit-content;
  flex-grow: 1;
  padding: 0 50px;
  position: relative;
}

.static-callouts-item:not(:last-child)::after {
  position: absolute;
  right: -12px;
  content: ' ';
  height: 100%;
  border-right: 4px solid $light-400;
  transform: skew(-12deg);
}

.static-callouts-item:first-child {
  padding-left: 0;
}

.static-callouts-item:last-child {
  padding-right: 0;
}

@include media-breakpoint-down(sm) {
  .static-callouts-container {
    flex-direction: column;
    row-gap: 1.5rem !important;
  }

  .static-callouts-item {
    padding: 0 !important;
  }

  .static-callouts-item::after {
    border: none !important;
  }
}

@include media-breakpoint-down(md) {
  .static-callouts-container-md {
    flex-wrap: wrap;
    row-gap: 3.125rem;
  }

  .static-callouts-item-md {
    padding: 0;
  }
  .static-callouts-item-md:nth-child(even) {
    padding-left: 3.125rem;
  }

  .static-callouts-item-md:nth-child(even)::after {
    border: none !important;
  }
}
