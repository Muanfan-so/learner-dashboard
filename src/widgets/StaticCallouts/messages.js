/* eslint-disable quotes */
import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  badgeTitle: {
    id: 'StaticCallouts.badgeTitle',
    defaultMessage: 'New',
    description: 'The title for item badge.',
  },
  heading: {
    id: 'StaticCallouts.heading',
    defaultMessage: 'More opportunities for you <markup>to learn</markup>',
    description: 'The heading for static callouts.',
  },
  subheading: {
    id: 'StaticCallouts.subheading',
    defaultMessage: `We've added 500+ learning opportunities to create one of the world's most comprehensive free-to-degree online learning platforms.`,
    description: 'The subheading for static callouts.',
  },
});

export default messages;
