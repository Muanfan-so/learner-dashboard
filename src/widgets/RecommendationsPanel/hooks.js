import React from 'react';

import { StrictDict } from 'utils';
import { RequestStates } from 'data/constants/requests';
import { handleEvent } from 'data/services/segment/utils';

import * as module from './hooks';
import api from './api';

export const searchCourseEventName = 'learner_home.widget.search_course';

export const state = StrictDict({
  requestState: (val) => React.useState(val), // eslint-disable-line
  data: (val) => React.useState(val), // eslint-disable-line
});

export const useFetchCourses = (setRequestState, setData) => {
  React.useEffect(() => {
    let isMounted = true;
    api.fetchRecommendedCourses().then((response) => {
      if (isMounted) {
        setRequestState(RequestStates.completed);
        setData(response);
      }
    }).catch(() => {
      if (isMounted) {
        setRequestState(RequestStates.failed);
      }
    });
    return () => { isMounted = false; };
    /* eslint-disable */
  }, []);
};

export const useRecommendationPanelData = () => {
  const [requestState, setRequestState] = module.state.requestState(RequestStates.pending);
  const [data, setData] = module.state.data({});
  module.useFetchCourses(setRequestState, setData);
  const courses = data.data?.courses || [];
  const isPersonalizedRecommendation = data.data?.isPersonalizedRecommendation || false;
  const courseSearchClickTracker = () => handleEvent(searchCourseEventName, {
    pageName: 'learner_home',
    linkType: 'button',
    linkCategory: 'search_button',
  });
  return {
    courses,
    isPersonalizedRecommendation,
    isLoaded: requestState === RequestStates.completed && courses.length > 0,
    isFailed: requestState === RequestStates.failed
      || (requestState === RequestStates.completed && courses.length === 0),
    isLoading: requestState === RequestStates.pending,
    courseSearchClickTracker,
  };
};

export default {
  useRecommendationPanelData,
};
