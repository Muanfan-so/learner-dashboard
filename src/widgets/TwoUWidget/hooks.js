import { useState, useEffect } from 'react';

import { breakpoints, useWindowSize } from '@edx/paragon';

import { StrictDict } from 'utils';
import { RequestStates } from 'data/constants/requests';
import { hooks as appHooks } from 'data/redux';

import { activateExperiment, isUserInVariation } from './experiment';
import * as module from './hooks';
import api from './api';

export const state = StrictDict({
  requestState: (val) => useState(val), // eslint-disable-line
  data: (val) => useState(val), // eslint-disable-line
});

const useIsMediumScreen = () => {
  const { width } = useWindowSize();
  return width >= breakpoints.medium.minWidth && width < breakpoints.medium.maxWidth;
};

const useOptimizelyExperiment = () => {
  const [show2ULobs, setShow2ULobs] = useState(false);
  const [isExperimentLoaded, setIsExperimentLoaded] = useState(false);

  const enterpriseData = appHooks.useEnterpriseDashboardData();
  const isEnterpriseUser = !!enterpriseData;

  useEffect(() => {
    activateExperiment(isEnterpriseUser);
    const timer = setTimeout(() => {
      const inVariation = isUserInVariation();
      setShow2ULobs(inVariation);
      setIsExperimentLoaded(true);
    }, 500);

    return () => clearTimeout(timer);
  }, [isEnterpriseUser]);

  return { show2ULobs, isExperimentLoaded };
};

export const useFetchTwoUWidgetContext = (setRequestState, setData) => {
  useEffect(() => {
    let isMounted = true;
    api.fetchTwoUWidgetContext().then((response) => {
      if (isMounted) {
        setRequestState(RequestStates.completed);
        setData(response.data);
      }
    }).catch(() => {
      if (isMounted) {
        setRequestState(RequestStates.failed);
      }
    });
    return () => { isMounted = false; };
    /* eslint-disable */
  }, []);
};

const useTwoUWidgetData = () => {
  const [requestState, setRequestState] = module.state.requestState(RequestStates.pending);
  const [data, setData] = module.state.data({});
  module.useFetchTwoUWidgetContext(setRequestState, setData);

  return {
    countryCode: data?.countryCode || "",
    isLoaded: requestState === RequestStates.completed,
    isFailed: requestState === RequestStates.failed,
    isLoading: requestState === RequestStates.pending,
  };
};

export { useIsMediumScreen, useOptimizelyExperiment, useTwoUWidgetData };
