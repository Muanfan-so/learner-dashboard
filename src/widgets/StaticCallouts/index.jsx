import React from 'react';
import classNames from 'classnames';

import { useIntl } from '@edx/frontend-platform/i18n';
import { Badge, Hyperlink, Icon } from '@edx/paragon';
import { ArrowForward } from '@edx/paragon/icons';

import { useCountryCode } from 'data/redux/hooks';
import useIsMediumScreen from './hooks';
import { getMerchandisingItems, countryCodeUS } from './constants';
import messages from './messages';
import './index.scss';

export const StaticCallouts = () => {
  const { formatMessage } = useIntl();
  const countryCode = useCountryCode();
  const merchandisingItems = getMerchandisingItems(countryCode);
  const isMediumScreen = useIsMediumScreen();
  const markup = { markup: children => <span className="d-inline-flex mb-0 text-brand-500">{children}</span> };
  const items = merchandisingItems.map(item => (
    <Hyperlink
      destination={item.itemURL}
      key={item.key}
      target="_blank"
      showLaunchIcon={false}
      className={classNames('static-callouts-item', {
        'static-callouts-item-md': isMediumScreen && countryCode === countryCodeUS,
      })}
    >
      <Badge variant="warning" className="item-badge mb-2">{formatMessage(messages.badgeTitle)}</Badge>
      <h3>
        {formatMessage(item.title)}
      </h3>
      <div className="card-description">
        {formatMessage(item.description)}
        <Icon className="d-inline-block align-middle mx-1" src={ArrowForward} />
      </div>
    </Hyperlink>
  ));

  return (
    <div className="callouts-container">
      <div>
        <h2>
          {formatMessage(messages.heading, markup)}
        </h2>
        <p className="mb-0 text-gray-700">
          {formatMessage(messages.subheading)}
        </p>
      </div>
      <div className={classNames('static-callouts-container', {
        'static-callouts-container-md': isMediumScreen && countryCode === countryCodeUS,
      })}
      >
        {items}
      </div>
    </div>
  );
};

export default StaticCallouts;
