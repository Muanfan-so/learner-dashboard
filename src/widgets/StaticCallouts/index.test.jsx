import React from 'react';
import { shallow, mount } from 'enzyme';

import { useCountryCode } from 'data/redux/hooks';
import StaticCallouts from '.';

jest.mock('data/redux/hooks', () => ({
  useCountryCode: jest.fn(),
}));

jest.mock('./hooks', () => ({
  __esModule: true,
  default: jest.fn(() => false),
}));

describe('StaticCallouts tests', () => {
  it('StaticCallouts snapshot', () => {
    expect(shallow(<StaticCallouts />)).toMatchSnapshot();
  });
  it('Display 4 static LOB cards for US users', () => {
    useCountryCode.mockReturnValueOnce('US');

    const wrapper = mount(<StaticCallouts />);
    expect(wrapper.find('div.static-callouts-container').children()).toHaveLength(4);
  });
  it('Display 3 static LOB cards for non-US users', () => {
    useCountryCode.mockReturnValueOnce('PK');

    const wrapper = mount(<StaticCallouts />);
    expect(wrapper.find('div.static-callouts-container').children()).toHaveLength(3);
  });
});
