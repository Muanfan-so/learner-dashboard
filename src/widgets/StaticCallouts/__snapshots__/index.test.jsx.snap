// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StaticCallouts tests StaticCallouts snapshot 1`] = `
<div
  className="callouts-container"
>
  <div>
    <h2>
      <format-message-function
        message={
          Object {
            "defaultMessage": "More opportunities for you <markup>to learn</markup>",
            "description": "The heading for static callouts.",
            "id": "StaticCallouts.heading",
          }
        }
        values={
          Object {
            "markup": [Function],
          }
        }
      />
    </h2>
    <p
      className="mb-0 text-gray-700"
    >
      We've added 500+ learning opportunities to create one of the world's most comprehensive free-to-degree online learning platforms.
    </p>
  </div>
  <div
    className="static-callouts-container"
  >
    <Hyperlink
      className="static-callouts-item"
      destination="https://www.edx.org/executive-education?vanguards_click=execed"
      key="execed"
      showLaunchIcon={false}
      target="_blank"
    >
      <Badge
        className="item-badge mb-2"
        variant="warning"
      >
        New
      </Badge>
      <h3>
        Executive Education
      </h3>
      <div
        className="card-description"
      >
        Short courses to develop leadership skills
        <Icon
          className="d-inline-block align-middle mx-1"
        />
      </div>
    </Hyperlink>
    <Hyperlink
      className="static-callouts-item"
      destination="https://www.edx.org/masters?vanguards_click=masters"
      key="masters"
      showLaunchIcon={false}
      target="_blank"
    >
      <Badge
        className="item-badge mb-2"
        variant="warning"
      >
        New
      </Badge>
      <h3>
        Master's Degrees
      </h3>
      <div
        className="card-description"
      >
        Online degree programs from top universities
        <Icon
          className="d-inline-block align-middle mx-1"
        />
      </div>
    </Hyperlink>
    <Hyperlink
      className="static-callouts-item"
      destination="https://www.edx.org/bachelors?vanguards_click=bachelors"
      key="bachelors"
      showLaunchIcon={false}
      target="_blank"
    >
      <Badge
        className="item-badge mb-2"
        variant="warning"
      >
        New
      </Badge>
      <h3>
        Bachelor's Degrees
      </h3>
      <div
        className="card-description"
      >
        Begin or complete a degree; fully online
        <Icon
          className="d-inline-block align-middle mx-1"
        />
      </div>
    </Hyperlink>
  </div>
</div>
`;
